src/OpenTelemetry/Internal/SelfDiagnosticsConfigRefresher.cs

```diff
@@ -41,24 +80,163 @@ internal class SelfDiagnosticsConfigRefresher : IDisposable
+            {
+                stream = null;
+                availableByteCount = 0;
+                return false;
+            }
+
+            try
+            {
+                var cachedViewStream = this.viewStream.Value;
+
+                // Each thread has its own MemoryMappedViewStream created from the only one MemoryMappedFile.
+                // Once worker thread updates the MemoryMappedFile, all the cached ViewStream objects become
+                // obsolete.
+                // Each thread creates a new MemoryMappedViewStream the next time it tries to retrieve it.
+                // Whether the MemoryMappedViewStream is obsolete is determined by comparing the current
+                // MemoryMappedFile object with the MemoryMappedFile object cached at the creation time of the
+                // MemoryMappedViewStream.
+                if (cachedViewStream == null || this.memoryMappedFileCache.Value != this.memoryMappedFile)
+                {
+                    // Race condition: The code might reach here right after the worker thread sets memoryMappedFile
+                    // to null in CloseLogFile().
+                    // In this case, let the NullReferenceException be caught and fail silently.
+                    // By design, all events captured will be dropped during a configuration file refresh if
+                    // the file changed, regardless whether the file is deleted or updated.
+                    cachedViewStream = this.memoryMappedFile.CreateViewStream();
+                    this.viewStream.Value = cachedViewStream;
+                    this.memoryMappedFileCache.Value = this.memoryMappedFile;
+                }
+
+                long beginPosition, endPosition;
+                do
+                {
+                    beginPosition = this.logFilePosition;
+                    endPosition = beginPosition + byteCount;
+                    if (endPosition >= this.logFileSize)
+                    {
+                        endPosition %= this.logFileSize;
+                    }
+                }
+                while (beginPosition != Interlocked.CompareExchange(ref this.logFilePosition, endPosition, beginPosition));
+                availableByteCount = (int)(this.logFileSize - beginPosition);
+                cachedViewStream.Seek(beginPosition, SeekOrigin.Begin);
+                stream = cachedViewStream;
+                return true;
+            }
+            catch (Exception)
+            {
+                stream = null;
+                availableByteCount = 0;
+                return false;
+            }
         }
 
-        /// <inheritdoc/>
-        public void Dispose()
+        private async Task Worker(CancellationToken cancellationToken)
         {
-            this.Dispose(true);
-            GC.SuppressFinalize(this);
+            await Task.Delay(ConfigurationUpdatePeriodMilliSeconds, cancellationToken).ConfigureAwait(false);
```

Comment:
We're pulling every 3 seconds, would it be more efficient to use something like [FileSystemWatcher](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher?view=netcore-3.1) where it would signal us if the file/directory changed?

Reply:
It might be a bit more efficient while it introduces other problems:
1. platform compat, there are know issues that FileSystemWatcher could miss the update.
2. notification timing - if a file got updated 10 times in 2 seconds, the cost could be higher.
3. semantic difference - e.g. folder move, permission change (from no access to have access), low file file restore, junction, hardlink - they have different behavior.

Given this is a very low level tool (collecting internal logs for a telemetry SDK), my general thinking would be leaning towards low-tech - something that might not be the most elegant/performant/efficient, but as reliable as a rock.

Reply:
Cool I trust your knowledge and experience on that. Maybe we should increase the time though? 5 or 10 seconds? I know people will want it to reflect instantly when they make a change, but I'm worried about busy work and tying up a thread. Assuming most of the time (99.9999%) people won't be interested in the diagnostics.

Reply:
5, 7 or 10 would work. No strong preference here.

Reply:
Created a PR #1523 for the update.